Mathlib/Tactic/CategoryTheory/RotateIsos/Cancelable.lean

```diff
@@ -0,0 +1,158 @@
+- Duplicate entries should not cause undefined behaviour, but are still intented to be avoided.
+- No assumption on order should be made about entries with equal priorities.
+
+The helper functions `insertCancelableFactory` and `insertCancelableFactory'`
+take care of inserting functions at their right place, and should be the prefered
+way of registering ways to cancel morphisms.
+-/
+initialize cancelableFactoriesRef :
+    IO.Ref <| CancelM <| List <| (Expr → Expr → MetaM (Option Cancelable)) × Nat ←
+  IO.mkRef <| pure []
+
+/-- Helper function that registers a function `f : Expr → Expr → MetaM (Option Cancelable)`
+as a way to construct `Cancelable` structure.
+This function does not check that this
+does not produce a duplicate entry and caller is responsible for ensuring first that they
+are not duplicating an entry.
+See the docstring of `cancelableFactoriesRef` for an explanation on the intended behaviour
+of such functions. -/
+def insertCancelableFactory (f : Expr → Expr → MetaM (Option Cancelable)) (p : Nat) : IO Unit :=
+  cancelableFactoriesRef.modify fun current =>
+    return List.insertP (fun x => p ≥ x.2) (f, p) (← current)
+
+/-- Helper function that registers a function `f : Expr → Expr → CancelM (Option Cancelable)`
+as a way to construct `Cancelable` structure.
+This function does not check that this
+does not produce a duplicate entry and caller is responsible for ensuring first that they
+are not duplicating an entry.
+See the docstring of `cancelableFactoriesRef` for an explanation on the intended behaviour
+of such functions. -/
+def insertCancelableFactory' (f : Expr → Expr → CancelM (Option Cancelable)) (p : Nat) :
+      IO Unit :=
+  cancelableFactoriesRef.modify fun current => do
+    let c ← read
+    return List.insertP (fun x => p ≥ x.2) (fun e w => (f e w).run c, p) (← current)
+
+/-- Given an expression `e`, if `e` is an expression for a cancelable morphism, returns
+a `Cancelable` structure such that `e.expr` is the original expression, using the rules to do
+so that are currently in context. Otherwise, returns none. -/
+def getCancelable? (e : Expr) : MetaM (Option Cancelable) := do
+  getCancelable?Aux e <| (← cancelableFactoriesRef.get).bind (fun x => return x.map Prod.fst)
+
+open _root_.CategoryTheory in
+/-- Given an expression of type `f₁ ≫ ⋯ ≫ fₙ`or `f₁ ≪≫ ⋯ ≪≫ fₙ`,
+assumed to be either fully left-associated or right-associated
+(depending on the argument `rev_assoc`),
+build a list of the cancellable morphisms (with their cancellation data) starting from
+the leftmost or rightmost (depending on the argument `rev`) until we hit a non-cancellable term.
+The function also returns a flag that is set if all of the morphisms are cancellable. -/
+partial def getCancelables (e : Expr) (rev rev_assoc: Bool) : MetaM (List Cancelable × Bool) := do
+  match (← whnfR e).getAppFnArgs with
+  | (``CategoryStruct.comp, #[_, _, _, _, _, l, r])
+  | (``Iso.trans, #[_, _, _, _, _, l, r]) =>
```

Comment:
This is probably best left for an upgrade later on, but I wonder if there should be a tag `@[comp]` that says if something is a composition or not. Then at this step you could instead check if the function name is tagged with `comp` or not (and `l, r` could be accessed from the back on the list, so this could be insensitive to the number of arguments)